app/src/main/java/com/futurice/freesound/mvi/BaseViewModel.kt

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2017 Futurice GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.futurice.freesound.mvi
+
+import com.futurice.freesound.feature.common.scheduling.SchedulerProvider
+import com.jakewharton.rx.replayingShare
+import io.reactivex.Observable
+import io.reactivex.disposables.SerialDisposable
+import io.reactivex.processors.PublishProcessor
+import timber.log.Timber
+
+abstract class BaseViewModel<E, D, M, C>(
```

Comment:
Most likely, yes. I'm currently thinking that it' over parameterized. Perhaps only `E`, `D` and `M` are necessarily. If nothing else, they are the more important of the interactions, so perhaps it's possible to hide `C`.